manta-pay/src/crypto/ecc/arkworks.rs

```diff
@@ -406,6 +406,31 @@ where
     }
 }
 
+impl<C, CV, const N: usize> ecc::PreprocessedScalarMul<Compiler<C>, N> for GroupVar<C, CV> where
+    C: ProjectiveCurve,
+    CV: CurveVar<C, ConstraintField<C>> {
+    /// Performs the scalar multiplication against a pre-computed table.
+    ///
+    /// The pre-computed table is a list of power-of-two multiples of `scalar`, such that
+    /// `table[i] = scalar * 2^i`.
+    fn preprocessed_scalar_mul(table: &[Self; N], scalar: &Self::Scalar, compiler: &mut Compiler<C>) -> Self::Output {
+        // Computes the standard little-endian double-and-add algorithm
+        // (Algorithm 3.26, Guide to Elliptic Curve Cryptography)
+        // adapted from https://github.com/arkworks-rs/r1cs-std/blob/50ab8ee5ba8c09637044718eceac38e98e0ea67c/src/groups/mod.rs#L110-L139
+
+        let _ = compiler;
+        let mut result: CV = CV::zero();
+        let scalar_bits = scalar.0.to_bits_le().expect("Bit decomposition is not allowed to fail.");
+        debug_assert_eq!(scalar_bits.len(), N, "Scalar is expected to have N bits.");
+        for (bit, base) in scalar_bits.into_iter().zip(table.iter()){
+            // compute `self + 2^i * scalar`
+            let scalar_plus_base = result.clone() + base.0.clone();
+            result = bit.select(&scalar_plus_base, &result).expect("Conditional select is not allowed to fail. ");
+        }
+        Self(result, PhantomData)
+    }
+}
```

Comment:
The `N` here represents the compile-time fixed and known number of bits of a scalar for the particular curve. This trait should not be implemented over all `N` (I wanted to use an associated constant but it's unstable to use that with slices rn). We can improve the documentation to help make this clearer.

Reply:
To which specific value of `N` shall we implement? I can write a macro to do that

Reply:
It should be implemented for the number of bits in a scalar. It's exactly the "debug assert" that you do, but we know at compile-time that the table is the correct size.

Reply:
currently we cannot put associate constant to const generic. For example, we cannot implement `ecc::PreprocessedScalarMul<Compiler<C>, Self::Scalar::MODULUS_BITS>`. We might need to implement that for every individual curve. 

I have implemented this trait only for Bls12_381. I also wrote a macro to easily support other curves if we need that

Reply:
done